src/thunderbots/software/ai/navigator/path_planner/astar.h

```diff
@@ -0,0 +1,149 @@
+#pragma once
+
+// I apologize if your CLion slows to a crawl
+#include <boost/graph/astar_search.hpp>
+#include <boost/graph/grid_graph.hpp>
+
+#include "ai/world/field.h"
+#include "path_planner.h"
+namespace AStar
+{
+    // TODO: potentially move a lot of stuff out of this header
+
+    // number of graph vertices per metre
+    // TODO: make this a dynamic parameter or passed in the constructor
+    // to AStarGridGraph
+    static constexpr unsigned long GRID_VERTEX_DENSITY = 20;
+    // edge cost type
+    typedef double edge_cost_t;
+    // 2D grid graph type
+    typedef boost::grid_graph<2> GridGraph2D;
+    // vertex type - a 2D point on a grid
+    typedef GridGraph2D::vertex_descriptor GridVertex;
+
+    // a hash function for a grid vertex, necessary for all the maps
+    // that boost::astar_search uses
+    struct grid_vertex_hash : std::unary_function<GridVertex, std::size_t>
+    {
+        std::size_t operator()(const GridVertex &gp) const;
+    };
+
+    class AStarGridGraph
+    {
+       public:
+        AStarGridGraph() = delete;
+
+        /**
+         * Constructs a grid graph from a Field object.
+         *
+         * @param field the field to create a graph for
+         */
+        explicit AStarGridGraph(const Field &field);
```

Comment:
What about robots? I imagine that we don't want to make graph node's inside obstacles?

Reply:
I was going to use the edge weights and heuristic to deal with obstacles instead, to avoid constructing vertices at every tick and because I dont *think* boost::grid_graph lets us remove/insert nodes

Reply:
Interesting. Have you considered using an implicit graph instead? An implicit graph is generated on an "as-needed" basis, rather then generating the entire thing in advance. If we define the function that generates the neighbors for a given vertex, we could specify that no vertex's are generated within obstacles. This would probably help seriously reduce the memory footprint as well!

Checkout page 12 of this paper (basically an implicit graph implementation that can be used with A* in boost): http://cs.krisbeevers.com/research/astar_bgl04.pdf

Let me know what you think (might be more appropriate to open a `proj_a_star_nav` channel in slack)!

Reply:
The process of using an implicit graph with boost::astar_search seems rather complex, and will probably be pretty hard to maintain. I did find [this library](https://github.com/subh83/DOSL) that seems substantially easier to use - I could take some time to look into using this next week.

Reply:
Hmm, that's a good point. I suppose we could implement our own, but this is rapidly going out of the scope of just getting a basic working navigator done. I'd say look into it, but don't feel the need to spend a lot of time on it, let's push to get a basic working version in and speed it up later.